File: practical work/1semester/T4/quasi.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

const double E = exp(1.0);
const double EPS = 1e-12; 

vector<double> thomas_algorithm(const vector<double>& a, const vector<double>& b, 
                                 const vector<double>& c, const vector<double>& d) {
    int n = d.size();
    vector<double> alpha(n), beta(n), x(n);
    
    alpha[0] = -c[0] / b[0];
    beta[0] = d[0] / b[0];
    
    for (int i = 1; i < n - 1; i++) {
        double denom = b[i] + a[i] * alpha[i-1];
        alpha[i] = -c[i] / denom;
        beta[i] = (d[i] - a[i] * beta[i-1]) / denom;
    }
    
    x[n-1] = (d[n-1] - a[n-1] * beta[n-2]) / (b[n-1] + a[n-1] * alpha[n-2]);
    for (int i = n - 2; i >= 0; i--) {
        x[i] = alpha[i] * x[i+1] + beta[i];
    }
    return x;
}

inline double safe_exp(double x, double max_val = 50.0) {
    return exp(max(min(x, max_val), -max_val));
}

inline double clip(double x, double lo, double hi) {
    return max(lo, min(x, hi));
}

double f_original(double x, double y, double yp) {
    double log_x = abs(log(x)) < EPS ? EPS : log(x);
    double exp_term = safe_exp(yp);
    double under_sqrt = 1.0/(x*x) + E*y*y/log_x - y*exp_term;
    return sqrt(max(under_sqrt, EPS));
}

double f_y(double x, double y, double yp) {
    double log_x = abs(log(x)) < EPS ? EPS : log(x);
    double exp_term = safe_exp(yp);
    double under_sqrt = 1.0/(x*x) + E*y*y/log_x - y*exp_term;
    double denom = 2.0 * sqrt(max(under_sqrt, EPS));
    double numer = 2.0*E*y/log_x - exp_term;
    return clip(numer/denom, -1e10, 1e10);
}

double f_yp(double x, double y, double yp) {
    double log_x = abs(log(x)) < EPS ? EPS : log(x);
    double exp_term = safe_exp(yp);
    double under_sqrt = 1.0/(x*x) + E*y*y/log_x - y*exp_term;
    double denom = 2.0 * sqrt(max(under_sqrt, EPS));
    double numer = -y * exp_term;
    return clip(numer/denom, -1e10, 1e10);
}

double g_n(double x, double y, double yp) {
    return f_original(x,y,yp) - f_y(x,y,yp)*y - f_yp(x,y,yp)*yp;
}

vector<double> gradient(const vector<double>& y, double h) {
    int n = y.size();
    vector<double> g(n);
    g[0] = (y[1] - y[0]) / h;
    for (int i = 1; i < n-1; i++)
        g[i] = (y[i+1] - y[i-1]) / (2*h);
    g[n-1] = (y[n-1] - y[n-2]) / h;
    return g;
}

pair<vector<double>, vector<double>> quasilinear_solve(int N = 200, double tol = 1e-6, int max_iter = 10000) {
    double a_val = E, b_val = E*E;
    double y_a = E, y_b = 2*E*E;
    
    vector<double> x(N), y(N);
    double h = (b_val - a_val) / (N - 1);
    
    for (int i = 0; i < N; i++) {
        x[i] = a_val + i * h;
        y[i] = x[i] * log(x[i]);
    }
    
    for (int it = 0; it < max_iter; it++) {
        vector<double> yp = gradient(y, h);
        
        vector<double> A(N), B(N), C(N), D(N);
        A[0] = 0; B[0] = 1; C[0] = 0; D[0] = y_a;
        A[N-1] = 0; B[N-1] = 1; C[N-1] = 0; D[N-1] = y_b;
        
        for (int i = 1; i < N-1; i++) {
            double p = -f_yp(x[i], y[i], yp[i]);
            double q = -f_y(x[i], y[i], yp[i]);
            double r = g_n(x[i], y[i], yp[i]);
            
            A[i] = 1.0/(h*h) - p/(2*h);
            B[i] = -2.0/(h*h) + q;
            C[i] = 1.0/(h*h) + p/(2*h);
            D[i] = r;
        }
        
        vector<double> y_new = thomas_algorithm(A, B, C, D);
        
        double diff = 0;
        for (int i = 0; i < N; i++)
            diff = max(diff, abs(y_new[i] - y[i]));
        
        cout << "Iter " << it+1 << ": " << diff << endl;
        
        y = y_new;
        if (diff < tol) {
            cout << "Converged" << endl;
            break;
        }
    }
    return {x, y};
}

int main() {
    auto [x, y] = quasilinear_solve(200, 1e-10);
    return 0;
}

File: practical work/1semester/T4/quasi.py
import numpy as np
import matplotlib.pyplot as plt

# Метод прогонки (Thomas algorithm)
def thomas_algorithm(a, b, c, d):
    n = len(d)
    alpha = np.zeros(n)
    beta = np.zeros(n)
    x = np.zeros(n)

    # Прямой ход
    alpha[0] = -c[0] / b[0]
    beta[0] = d[0] / b[0]

    for i in range(1, n - 1):
        denom = b[i] + a[i] * alpha[i - 1]
        alpha[i] = -c[i] / denom
        beta[i] = (d[i] - a[i] * beta[i - 1]) / denom

    # Обратный ход
    x[n - 1] = (d[n - 1] - a[n - 1] * beta[n - 2]) / (b[n - 1] + a[n - 1] * alpha[n - 2])
    for i in range(n - 2, -1, -1):
        x[i] = alpha[i] * x[i + 1] + beta[i]

    return x

def safe_exp(x, max_val=50.0):
    return np.exp(np.clip(x, -max_val, max_val))

def f_original(x, y, y_prime):
    eps = 1e-12
    log_x = np.where(np.abs(np.log(x)) < eps, eps, np.log(x))
    exp_term = safe_exp(y_prime)
    under_sqrt = 1.0 / x**2 + np.e * y**2 / log_x - y * exp_term
    return np.sqrt(np.maximum(under_sqrt, eps))

def f_y(x, y, y_prime):
    eps = 1e-12
    log_x = np.where(np.abs(np.log(x)) < eps, eps, np.log(x))
    exp_term = safe_exp(y_prime)
    under_sqrt = 1.0 / x**2 + np.e * y**2 / log_x - y * exp_term
    denominator = 2.0 * np.sqrt(np.maximum(under_sqrt, eps))
    numerator = 2.0 * np.e * y / log_x - exp_term
    return np.clip(numerator / denominator, -1e10, 1e10)

def f_y_prime(x, y, y_prime):
    eps = 1e-12
    log_x = np.where(np.abs(np.log(x)) < eps, eps, np.log(x))
    exp_term = safe_exp(y_prime)
    under_sqrt = 1.0 / x**2 + np.e * y**2 / log_x - y * exp_term
    denominator = 2.0 * np.sqrt(np.maximum(under_sqrt, eps))
    numerator = -y * exp_term
    return np.clip(numerator / denominator, -1e10, 1e10)

def g_n(x, y, y_prime):
    return f_original(x, y, y_prime) - f_y(x, y, y_prime) * y - f_y_prime(x, y, y_prime) * y_prime

# Решение линейной задачи методом прогонки
def linear_bvp_solve_sweep(p, q, r, x, y_left, y_right):
    N = len(x)
    h = x[1] - x[0]
    a = np.zeros(N)
    b = np.zeros(N)
    c = np.zeros(N)
    d = np.zeros(N)

    a[0], b[0], c[0], d[0] = 0, 1, 0, y_left
    a[-1], b[-1], c[-1], d[-1] = 0, 1, 0, y_right

    for i in range(1, N - 1):
        a[i] = 1.0 / h**2 - p[i] / (2.0 * h)
        b[i] = -2.0 / h**2 + q[i]
        c[i] = 1.0 / h**2 + p[i] / (2.0 * h)
        d[i] = r[i]

    return thomas_algorithm(a, b, c, d)

# Метод квазилинеаризации
def quasilinear_solve(N=200, tol=1e-10, max_iter=10000):
    a, b = np.e, np.e**2
    y_a, y_b = np.e, 2 * np.e**2
    x = np.linspace(a, b, N)
    h = x[1] - x[0]

    # Начальное приближение
    y = x * np.log(x)

    for it in range(max_iter):
        y_prime = np.gradient(y, h)
        p = -f_y_prime(x, y, y_prime)
        q = -f_y(x, y, y_prime)
        r = g_n(x, y, y_prime)

        y_new = linear_bvp_solve_sweep(p, q, r, x, y_a, y_b)
        diff_norm = np.linalg.norm(y_new - y, np.inf)

        print(f"Итерация {it+1}: {diff_norm:.3e}")
        if diff_norm < tol:
            y = y_new
            print("Сходимость достигнута")
            break
        y = y_new

    return x, y

if __name__ == "__main__":
    x_result, y_result = quasilinear_solve(N=200, tol=1e-6)

    plt.figure(figsize=(10, 6))
    plt.plot(x_result, y_result, 'b-', linewidth=2, label='Решение')
    plt.plot(x_result, x_result * np.log(x_result), 'r--', linewidth=1, label='Начальное приближение')
    plt.xlabel('x')
    plt.ylabel('y(x)')
    plt.title('Квазилинеаризация')
    plt.legend()
    plt.grid(True)
    plt.tight_layout()
    plt.show()
